File: packages/elixir/lib/kreuzberg/code_export_info.ex
defmodule Kreuzberg.CodeExportInfo do
  @moduledoc """
  Export statement information.

  ## Fields

    * `:name` - Exported item name
    * `:kind` - Export kind
    * `:span` - Source span
  """

  @type t :: %__MODULE__{
          name: String.t(),
          kind: String.t(),
          span: Kreuzberg.CodeSpan.t()
        }

  defstruct name: "",
            kind: "",
            span: %Kreuzberg.CodeSpan{}

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    span =
      case Map.get(data, "span", Map.get(data, :span)) do
        %Kreuzberg.CodeSpan{} = s -> s
        map when is_map(map) -> Kreuzberg.CodeSpan.from_map(map)
        _ -> %Kreuzberg.CodeSpan{}
      end

    %__MODULE__{
      name: Map.get(data, "name", Map.get(data, :name, "")),
      kind: Map.get(data, "kind", Map.get(data, :kind, "")),
      span: span
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = info) do
    %{
      "name" => info.name,
      "kind" => info.kind,
      "span" => Kreuzberg.CodeSpan.to_map(info.span)
    }
  end
end

File: packages/php/src/Types/CodeExportInfo.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Export statement information.
 *
 * @property-read string $name Exported item name
 * @property-read string $kind Export kind
 * @property-read CodeSpan $span Source span
 */
readonly class CodeExportInfo
{
    public function __construct(
        public string $name,
        public string $kind,
        public CodeSpan $span,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $name */
        $name = $data['name'] ?? '';

        /** @var string $kind */
        $kind = $data['kind'] ?? '';

        /** @var array<string, mixed> $spanData */
        $spanData = $data['span'] ?? [];

        return new self(
            name: $name,
            kind: $kind,
            span: CodeSpan::fromArray($spanData),
        );
    }
}
